src/librustc_metadata/creader.rs

```diff
@@ -839,6 +879,46 @@ impl<'a> CrateLoader<'a> {
         });
     }
 
+    fn report_unused_deps(&mut self, krate: &ast::Crate) {
+        // Get set of specified crates on the command line
+        let mut externs = BTreeSet::default();
+        for (name, _ext) in self.sess.opts.externs.iter() {
+            externs.insert(Symbol::intern(name));
+        }
+
+        // Remove all resolved crates which this crate references directly
+        self.cstore.iter_crate_data_names(|_, crate_data, names| {
+            info!(
+                "resolved crate `{}` names {:?} direct_extern_crate {}",
+                crate_data.name(),
+                names,
+                crate_data.direct_extern_crate()
+            );
+            if crate_data.direct_extern_crate() {
+                for name in names {
+                    externs.remove(name);
+                }
+            }
+        });
+
+        // Make a point span rather than covering the whole file
+        let spandata = krate.span.data();
+        let span = spandata.with_hi(spandata.lo);
+        // Complain about anything left over
+        for ext in externs {
+            self.sess.parse_sess.buffer_lint(
+                lint::builtin::UNUSED_CRATE_DEPS,
+                span,
```

Comment:
```suggestion
                DUMMY_SP,
```
Dummy span actually points to `main_file.rs:0:0` so using the crate span should be unnecessary.

Reply:
I tried this initially, but using it suppresses any kind of location reporting at all:
```
warning: External crate `bar` unused in `warn_cmdline`. Remove the dependency or add `use bar as _;`.
   |
   = note: requested on the command line with `-W unused-crate-deps`

warning: 1 warning emitted

```